polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -298,6 +324,154 @@ async fn run<Context>(
 	}
 }
 
+// Returns the new session index if it is greater than the current one.
+async fn new_session_index<Sender>(
+	sender: &mut Sender,
+	session_index: Option<SessionIndex>,
+	relay_parent: Hash,
+) -> Option<SessionIndex>
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	let Ok(Ok(new_session_index)) =
+		util::request_session_index_for_child(relay_parent, sender).await.await
+	else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch session index from runtime API",
+		);
+		return None
+	};
+
+	session_index.map_or(Some(new_session_index), |index| {
+		if new_session_index > index {
+			Some(new_session_index)
+		} else {
+			None
+		}
+	})
+}
+
+// Returns true if the node is an authority in the next session.
+async fn check_next_session_authority<Sender>(
+	sender: &mut Sender,
+	keystore: KeystorePtr,
+	relay_parent: Hash,
+) -> bool
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	// In spite of function name here we request past, present and future authorities.
+	// It's ok to stil prepare PVFs in other cases, but better to request only future ones.
+	let Ok(Ok(authorities)) = util::request_authorities(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch authorities from runtime API",
+		);
+		return false
+	};
+
+	authorities
+		.iter()
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]))
+}
+
+// Sends PVF with unknown code hashes to the validation host returning the list of code hashes sent.
+async fn prepare_pvfs_for_current_candidates<Sender>(
+	sender: &mut Sender,
+	mut validation_backend: impl ValidationBackend,
+	relay_parent: Hash,
+	already_prepared: &HashSet<ValidationCodeHash>,
+) -> Option<Vec<ValidationCodeHash>>
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	let Ok(Ok(events)) = util::request_candidate_events(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch candidate events from runtime API",
+		);
+		return None
+	};
+	let code_hashes = events
+		.into_iter()
+		.filter_map(|e| match e {
+			CandidateEvent::CandidateBacked(receipt, ..) => {
+				let h = receipt.descriptor.validation_code_hash;
+				if already_prepared.contains(&h) {
+					None
+				} else {
+					Some(h)
+				}
+			},
+			_ => None,
+		})
+		.collect::<Vec<_>>();
+
+	let Ok(executor_params) = util::executor_params_at_relay_parent(relay_parent, sender).await
+	else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch executor params for the session",
+		);
+		return None
+	};
+	let timeout = pvf_prep_timeout(&executor_params, PvfPrepKind::Precheck);
```

Comment:
```suggestion
	let timeout = pvf_prep_timeout(&executor_params, PvfPrepKind::Prepare);
```

We're preparing for execution, so the shorter timeout must be used.